src/retime.jl

```diff
@@ -0,0 +1,233 @@
+struct Max <: AggregationMethod end
+struct Count <: AggregationMethod end
+struct Sum <: AggregationMethod end
+struct Median <: AggregationMethod end
+struct First <: AggregationMethod end
+struct Last <: AggregationMethod end
+
+# Extrapolation methods
+struct FillConstant{V} <: ExtrapolationMethod
+    value::V
+end
+struct NearestExtrapolate <: ExtrapolationMethod end
+struct MissingExtrapolate <: ExtrapolationMethod end
+struct NaNExtrapolate <: ExtrapolationMethod end
+
+_toInterpolationMethod(x::Symbol) = _toInterpolationMethod(Val(x))
+_toInterpolationMethod(::Val{:linear}) = Linear()
+_toInterpolationMethod(::Val{:previous}) = Previous()
+_toInterpolationMethod(::Val{:next}) = Next()
+_toInterpolationMethod(::Val{:nearest}) = Nearest()
+_toInterpolationMethod(x::InterpolationMethod) = x
+
+_toAggregationMethod(x::Symbol) = _toAggregationMethod(Val(x))
+_toAggregationMethod(::Val{:mean}) = Mean()
+_toAggregationMethod(::Val{:min}) = Min()
+_toAggregationMethod(::Val{:max}) = Max()
+_toAggregationMethod(::Val{:count}) = Count()
+_toAggregationMethod(::Val{:sum}) = Sum()
+_toAggregationMethod(::Val{:median}) = Median()
+_toAggregationMethod(::Val{:first}) = First()
+_toAggregationMethod(::Val{:last}) = Last()
+_toAggregationMethod(x::AggregationMethod) = x
+
+_toExtrapolationMethod(x::Symbol) = _toExtrapolationMethod(Val(x))
+_toExtrapolationMethod(::Val{:fillconstant}) = FillConstant(0.0)
+_toExtrapolationMethod(::Val{:nearest}) = NearestExtrapolate()
+_toExtrapolationMethod(::Val{:missing}) = MissingExtrapolate()
+_toExtrapolationMethod(::Val{:nan}) = NaNExtrapolate()
+_toExtrapolationMethod(x::ExtrapolationMethod) = x
+
+function retime(ta, new_dt::Dates.Period; kwargs...)
+    new_timestamps = timestamp(ta)[1]:new_dt:timestamp(ta)[end]
+    return retime(ta, new_timestamps; kwargs...)
+end
+
+function retime(ta, period::Function; kwargs...)
+    new_timestamps = map(i -> first(i), _split(timestamp(ta), period))
+    return retime(ta, new_timestamps; kwargs...)
+end
+
+function retime(
+    ta::TimeArray{T,N,D,A},
+    new_timestamps::AbstractVector{DN};
+    upsample::Union{Symbol,InterpolationMethod}=Previous(),
+    downsample::Union{Symbol,AggregationMethod}=Mean(),
+    extrapolate::Union{Symbol,ExtrapolationMethod}=NearestExtrapolate(),
+    skip_missing::Bool=true,
+) where {T,N,D,A,DN}
+    upsample = _toInterpolationMethod(upsample)
+    downsample = _toAggregationMethod(downsample)
+    extrapolate = _toExtrapolationMethod(extrapolate)
+
+    new_values = __get_new_values(
+        T, length(new_timestamps), size(values(ta), 2), extrapolate, skip_missing
+    )
+    old_timestamps = convert(Vector{DN}, timestamp(ta))
+    old_values = values(ta)
+    @views begin
+        for col_i in 1:size(old_values, 2)
+            if skip_missing
+                idx = findall(x -> !ismissing(x) && !isnan(x), old_values[:, col_i])
+            else
+                idx = ones(Int, length(old_timestamps))
```

Comment:
fine to have this? I don't run the test against this patch yet

```suggestion
                idx = Colon()
```